### PR TITLE
[Snyk] Security upgrade torch from 2.5.0 to 2.11.0

### PR DESCRIPTION
**Pull request by karly-quick**
snyk-top-banner

#### Snyk changed the following file(s):

- `runtime/tools/ttrt/requirements.txt`

#### Breaking Change Risk

Merge Risk: High

> **Notice:** This assessment is enhanced by AI.

---

> [!IMPORTANT]
>
> - Check the changes in this PR to ensure they won't cause issues with your project.
> - Max score is 1000. Note that the real score may have changed since the PR was raised.
> - This PR was automatically created by Snyk using the credentials of a real user.
> - Some vulnerabilities couldn't be fully fixed and so Snyk will still find them when the project is tested again. This may be because the vulnerability existed within more than one direct dependency, but not all of the affected dependencies could be upgraded.

---

**Note:** _You are seeing this because you or someone else with access to this repository has authorized Snyk to open fix PRs._

For more information: 
🧐 [View latest project report](https://app.snyk.io/org/tt-test-rust/project/73fb8371-3c86-4cdd-999c-15dc1b102a9c?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr)
📜 [Customise PR templates](https://docs.snyk.io/scan-using-snyk/pull-requests/snyk-fix-pull-or-merge-requests/customize-pr-templates?utm_source=github&utm_content=fix-pr-template)
🛠 [Adjust project settings](https://app.snyk.io/org/tt-test-rust/project/73fb8371-3c86-4cdd-999c-15dc1b102a9c?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr/settings)
📚 [Read about Snyk's upgrade logic](https://docs.snyk.io/scan-with-snyk/snyk-open-source/manage-vulnerabilities/upgrade-package-versions-to-fix-vulnerabilities?utm_source=github&utm_content=fix-pr-template)

---

**Learn how to fix vulnerabilities with free interactive lessons:**

🦉 [Out-of-bounds Write](https://learn.snyk.io/lesson/out-of-bounds-write/?source&#x3D;registry-oss-fix-pr&amp;loc&#x3D;fix-pr)

[//]: # 'snyk:metadata:{"breakingChangeRiskLevel":"high","FF_showPullRequestBreakingChanges":true,"FF_showPullRequestBreakingChangesWebSearch":false,"customTemplate":{"variablesUsed":[],"fieldsUsed":[]},"dependencies":[{"name":"torch","from":"2.5.0","to":"2.11.0"}],"env":"prod","issuesToFix":["SNYK-PYTHON-TORCH-10337828"],"prId":"5177969d-f025-4614-97f8-45891324a44e","prPublicId":"5177969d-f025-4614-97f8-45891324a44e","packageManager":"pip","priorityScoreList":[150],"projectPublicId":"73fb8371-3c86-4cdd-999c-15dc1b102a9c","projectUrl":"https://app.snyk.io/org/tt-test-rust/project/73fb8371-3c86-4cdd-999c-15dc1b102a9c?utm_source=github&utm_medium=referral&page=fix-pr","prType":"fix","templateFieldSources":{"branchName":"default","commitMessage":"default","description":"default","title":"default"},"templateVariants":["updated-fix-title","priorityScore"],"type":"auto","upgrade":[],"vulns":["SNYK-PYTHON-TORCH-10337828"],"patch":[],"isBreakingChange":false,"remediationStrategy":"vuln"}'

### CODE REVIEW

**Comment by karly-quick** [comment]
Merge Risk: High

The upgrade from PyTorch 2.5.0 to 2.11.0 spans multiple minor releases and introduces several significant breaking changes, deprecations, and environment updates. This is a high-risk upgrade requiring careful code and environment validation.  
  
**Key Breaking Changes & Deprecations:**  
  
- **`torch.load` Security Change (from v2.6):** The `weights_only` parameter in `torch.load` now defaults to `True` for security reasons. If you are loading model checkpoints that contain anything other than weights (e.g., pickled code), you must explicitly set `weights_only=False` to avoid errors.  
  
- **TorchScript Deprecation (from v2.10):** TorchScript is now deprecated and users are strongly encouraged to migrate to `torch.export` for model serialization and deployment. Future releases will remove TorchScript, so this migration is critical for long-term support.  
  
- **ONNX Exporter Default Change (from v2.10):** The default ONNX exporter has been switched to the `dynamo` backend. This may affect the structure of exported models. To use the old behavior, you must manually specify the legacy exporter.  
  
- **Python Version Support:** Support for Python 3.9 was dropped in version 2.9. You must be using Python 3.10 or newer.  
  
**Environment & Hardware Compatibility:**  
  
- **Default CUDA Version:** The default pre-built binaries for `pip` now ship with CUDA 13.0 as of PyTorch 2.11. Builds for other CUDA versions like 12.6 are still available but may require specifying a different index URL during installation.  
  
- **GPU Architecture Support:** Support for older NVIDIA GPUs has been removed in newer CUDA builds. Notably:  
  - **Volta (e.g., V100):** Support is removed from CUDA 12.8 and 12.9 builds starting with PyTorch 2.11. Users with Volta GPUs must use the CUDA 12.6 build or compile from source.  
  - **Maxwell & Pascal:** Support was removed in PyTorch 2.8 for CUDA 12.8+ builds.  
  
- **Build from Source:** PyTorch 2.11 now requires a C++20 compatible compiler to build from source.  
  
**Recommendation:**  
  
1.  **Review `torch.load` usage:** Audit all calls to `torch.load` and explicitly set `weights_only=False` where necessary.  
2.  **Plan TorchScript Migration:** If you use TorchScript for production models, begin migrating to `torch.export` immediately.  
3.  **Validate Environment:** Ensure your Python version (&gt;=3.10) and CUDA toolkit/driver versions are compatible with the new release and your hardware. Pay close attention to GPU compatibility if using older models.  
4.  **Test ONNX Exports:** If you export models to ONNX, re-validate your export process and the resulting models due to the new `dynamo` backend default.  
  
Source: [PyTorch Release Notes](https://github.com/pytorch/pytorch/releases)

> **Notice 🤖:** This content was augmented using **artificial intelligence**. AI-generated content **may** contain **errors** and should be **reviewed** for accuracy before use.

[//]: # (snyk:breakability:{"publicId":"f15f552a-6ca1-47c6-93fc-c0e91cfa891f"})

**Comment by karly-quick** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/tt-test-rust/pr-checks/72ab48b7-199e-44f5-920e-500dade285a4?refs=2fd0c1b5fe1218ac58b735693d75152c01ff26a3&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/tt-test-rust/pr-checks/72ab48b7-199e-44f5-920e-500dade285a4/license?refs=2fd0c1b5fe1218ac58b735693d75152c01ff26a3&source=prChecksComment) |
| :white_check_mark: | Code Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/tt-test-rust/project/92274670-47b9-4055-9066-adf190b1ff4d/pr-check/13e21049-c732-4c2a-92c4-0f39967f3e46?refs=2fd0c1b5fe1218ac58b735693d75152c01ff26a3&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).
